torch/testing/asserts.py

```diff
@@ -0,0 +1,186 @@
+    change this behavior with the ``strict_attributes`` parameter.
+
+    Depending on the strictness some additional steps may be performed before the value comparison:
+
+    - If the tensors don't live in the same device memory they are moved to CPU memory.
+    - If the tensors don't have the same ``dtype``, they are converted to the one determined by
+        :func:`torch.promote_types`.
+
+    Args:
+        tensor1 (torch.Tensor): First tensor.
+        tensor2 (torch.Tensor): Second tensor.
+        warn_floating_point (bool): Emit a warning if at least one of the inputs is a floating point tensor.
+        **strict_attributes (bool): Set strictness for attribute equality checking. If omitted, this defaults to
+            ``shape=True, dtype=True``.
+
+    Raises:
+        AssertionError: If the specified attributes do not match.
+        AssertionError: If the tensors are not equal.
+    """
+    is_floating_point = tensor1.dtype.is_floating_point or tensor2.dtype.is_floating_point
+    if is_floating_point and warn_floating_point:
+        warnings.warn(
+            "Due to the limitations of floating point arithmetic, comparing floating-point tensors for equality is not "
+            "recommended. Use the respective 'allclose' variant instead.",
+            UsageWarning,
+        )
+
+    _compare_tensors(tensor1, tensor2, _tensors_equal_asserter, **strict_attributes)
+
+
+def _tensors_allclose_asserter(
+    tensor1: torch.Tensor, tensor2: torch.Tensor, rtol: Optional[float], atol: Optional[float]
+) -> None:
+    default_rtol, default_atol = _get_default_rtol_and_atol(tensor1, tensor2)
+    rtol = rtol if rtol is not None else default_rtol
+    atol = atol if atol is not None else default_atol
+
+    if not torch.allclose(tensor1, tensor2, rtol=rtol, atol=atol):
+        raise AssertionError("ADDME")
+
+
+@_hide_internal_traceback
+def assert_tensors_allclose(
+    tensor1: torch.Tensor,
+    tensor2: torch.Tensor,
+    *,
+    rtol: Optional[float] = None,
+    atol: Optional[float] = None,
+    **strict_attributes: bool,
+) -> None:
+    """Assert that two tensors are equal up to a desired precision.
+
+    Before the values are compared, the ``shape`` and the ``dtype`` of the tensors are checked for equality. You can
+    change this behavior with the ``strict_attributes`` parameter.
+
+    Depending on the strictness some additional steps may be performed before the value comparison:
+
+    - If the tensors don't live in the same device memory they are moved to CPU memory.
+    - If the tensors don't have the same ``dtype``, they are converted to the one determined by
+        :func:`torch.promote_types`.
+
+    Args:
+        tensor1 (torch.Tensor): First tensor.
+        tensor2 (torch.Tensor): Second tensor.
+        rtol (Optional[float]): Relative tolerance. If omitted, a default value is determined by the ``dtype`` of the
```

Comment:
"determined is vague." We should document the lookup in the operator description.

Reply:
I think we shouldn't expose the function (and with it some documentation), but I can simply add a table to this docstring.

Reply:
Table in this docstring sounds perfect.